File: backend/app/schemas/notification.py
"""Notification API schemas."""

from __future__ import annotations

import uuid
from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field

from app.models.enums import (
    NotificationChannel,
    NotificationDeliveryStatus,
    NotificationType,
)


class NotificationResponse(BaseModel):
    id: uuid.UUID
    user_id: uuid.UUID
    title: str
    message: str
    notification_type: NotificationType
    channel: NotificationChannel
    delivery_status: NotificationDeliveryStatus
    is_read: bool
    read_at: datetime | None = None
    link: str | None = None
    meta: dict[str, Any] | None = None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class NotificationListResponse(BaseModel):
    items: list[NotificationResponse]
    total: int
    page: int
    page_size: int
    pages: int
    unread_count: int = 0
    channel_counts: dict[str, int] = Field(default_factory=dict)


class UnreadCountResponse(BaseModel):
    unread_count: int


class MarkReadRequest(BaseModel):
    is_read: bool = True


class NotificationPreferenceResponse(BaseModel):
    email_enabled: bool
    in_app_enabled: bool
    push_enabled: bool
    vapid_public_key: str | None = None
    smtp_configured: bool = False
    push_configured: bool = False

    model_config = {"from_attributes": True}


class NotificationPreferenceUpdate(BaseModel):
    email_enabled: bool | None = None
    in_app_enabled: bool | None = None
    push_enabled: bool | None = None


class PushSubscribeRequest(BaseModel):
    endpoint: str = Field(..., min_length=10)
    p256dh: str = Field(..., min_length=10)
    auth: str = Field(..., min_length=8)
    user_agent: str | None = Field(default=None, max_length=500)


class PushUnsubscribeRequest(BaseModel):
    endpoint: str = Field(..., min_length=10)


class PushSubscriptionResponse(BaseModel):
    id: uuid.UUID
    endpoint: str
    created_at: datetime

    model_config = {"from_attributes": True}


class NotifyTestRequest(BaseModel):
    title: str = Field(default="Test notification", max_length=255)
    message: str = Field(default="This is a test from HirePulse.", max_length=2000)
    channels: list[NotificationChannel] | None = None

File: frontend/src/types/notification.ts
export type NotificationType = "info" | "success" | "warning" | "alert"
export type NotificationChannel = "in_app" | "email" | "sms" | "push"
export type NotificationDeliveryStatus =
  | "pending"
  | "sent"
  | "failed"
  | "skipped"

export type AppNotification = {
  id: string
  user_id: string
  title: string
  message: string
  notification_type: NotificationType
  channel: NotificationChannel
  delivery_status: NotificationDeliveryStatus
  is_read: boolean
  read_at: string | null
  link: string | null
  meta: Record<string, unknown> | null
  created_at: string
  updated_at: string
}

export type NotificationListResponse = {
  items: AppNotification[]
  total: number
  page: number
  page_size: number
  pages: number
  unread_count: number
  channel_counts: Record<string, number>
}

export type UnreadCountResponse = {
  unread_count: number
}

export type NotificationPreferences = {
  email_enabled: boolean
  in_app_enabled: boolean
  push_enabled: boolean
  vapid_public_key: string | null
  smtp_configured: boolean
  push_configured: boolean
}

export type NotificationPreferenceUpdate = {
  email_enabled?: boolean
  in_app_enabled?: boolean
  push_enabled?: boolean
}

export type NotificationListParams = {
  page?: number
  page_size?: number
  channel?: NotificationChannel
  unread_only?: boolean
}
